plugins/wacom-usb/fu-wac-plugin.c

```diff
@@ -33,6 +33,41 @@ fu_wac_plugin_write_firmware(FuPlugin *plugin,
 	return fu_device_write_firmware(device, blob_fw, progress, flags, error);
 }
 
+static gboolean
+fu_wac_plugin_composite_prepare(FuPlugin *self, GPtrArray *devices, GError **error)
+{
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device = g_ptr_array_index(devices, i);
+		if (FU_IS_WAC_DEVICE(device)) {
+			g_autoptr(FuDeviceLocker) locker = fu_device_locker_new(device, error);
+			if (locker == NULL)
+				return FALSE;
+			g_info("switching main device to flash loader");
+			if (!fu_wac_device_switch_to_flash_loader(FU_WAC_DEVICE(device), error))
+				return FALSE;
```

Comment:
If this loop has to terminate early because of an error, will the cleanup function be called to return them back to normal operation?

Reply:
If composite_prepare() fails, then we don't call composite_cleanup(). If anything in the write process fails then we **do** call composite_cleanup() -- i.e. we recover the device back to runtime mode. I don't know if it makes sense to call composite_cleanup() if composite_prepare() fails -- @superm1 -- what would you expect?

Reply:
I think I was misunderstanding what devices would be encountered over a single execution of this loop (see my other comments below). If only one FuWacDevice is contained in the list of devices that are looped over, then cleanup should indeed be unnecessary for us.